### PR TITLE
🌱 Add logging guidelines

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:

This PR documents the logging guidelines we are using in CAPI - pending merge of https://github.com/kubernetes-sigs/cluster-api/pull/6150

### CODE REVIEW

File: docs/book/src/developer/logging.md
Comment (chrischdi) [inline docs/book/src/developer/logging.md:73]:
```suggestion
- The logger passed to each reconcile call has a unique `reconcileID`, so all the logs being written during a single 
```

**Review by chrischdi — COMMENTED** [review]
small typo.

Great to read and understand 👍

File: REVIEWING.md
Comment (killianmuldoon) [inline REVIEWING.md:124]:
```suggestion
and all the required info is available when troubleshooting issues.
```

File: docs/book/src/developer/logging.md
Comment (killianmuldoon) [inline docs/book/src/developer/logging.md:70]:
```suggestion
Contextual logging is also embedded in controller runtime; In Cluster API we use contextual logging via controller runtime's
```

File: docs/book/src/developer/logging.md
Comment (killianmuldoon) [inline docs/book/src/developer/logging.md:40]:
```suggestion
Starting from the above foundations, then the long tail of small improvements will consist of following activities: 
```

File: docs/book/src/developer/logging.md
Comment (killianmuldoon) [inline docs/book/src/developer/logging.md:35]:
```suggestion
- Support for structured logging in all the Cluster API controllers (see [log format](#log-format)).
```

**Review by killianmuldoon — COMMENTED** [review]
A couple of tiny nits - but looks good to me otherwise!

**Comment by fabriziopandini** [comment]
/hold 
for CAPI office hours + lazy consensus

File: REVIEWING.md
Comment (vincepri) [inline REVIEWING.md:124]:
```suggestion
While developing controllers in Cluster API a key requirement is to add logging to observe the system and 
to help troubleshooting issues.
```

File: REVIEWING.md
Comment (vincepri) [inline REVIEWING.md:126]: ?

File: docs/book/src/developer/logging.md
Comment (vincepri) [inline docs/book/src/developer/logging.md:7]: We should define `users` in this context, SRE & developers can be users at the same time

Comment (vincepri) [reply docs/book/src/developer/logging.md:7]: Would you want to say application developer? Who's this persona?

File: docs/book/src/developer/logging.md
Comment (vincepri) [inline docs/book/src/developer/logging.md:13]:
```suggestion
- **Cluster API developers MUST log!**
```

File: docs/book/src/developer/logging.md
Comment (vincepri) [inline docs/book/src/developer/logging.md:23]: Remove this? It's a bit hard to follow

File: docs/book/src/developer/logging.md
Comment (vincepri) [inline docs/book/src/developer/logging.md:31]: Remove these? Let's get right to the point of defining our logging

File: docs/book/src/developer/logging.md
Comment (vincepri) [inline docs/book/src/developer/logging.md:133]: Debatable? The limit of verbosity should be in the `--v` flag level, rather than limiting ourselves from the get-go. For example, a very high level might mean that we can allow tracing in some cases, which is close to an IDE debugger. You can't necessarily attach a debugger in production systems

File: REVIEWING.md
Comment (killianmuldoon) [reply REVIEWING.md:126]: Can we just force merge with the correct link? This is pretty bad UX for adding pages to the book!

File: docs/book/src/developer/logging.md
Comment (CecileRobertMichon) [inline docs/book/src/developer/logging.md:155]:
```suggestion
we encourage providers to adopt and contribute to the guidelines defined in this document.
```

File: REVIEWING.md
Comment (fabriziopandini) [reply REVIEWING.md:126]: REVIEWING.md is an edge case, because this page is defined outside of the book but also imported in the book.
So, in order to make this page to work from outside the book you need to use the generated url of the target page https://cluster-api.sigs.k8s.io/developer/logging.html, that will exists only when the book is merged.

Comment (fabriziopandini) [reply REVIEWING.md:126]: see https://github.com/kubernetes-sigs/cluster-api/pull/6969#discussion_r928828230

File: docs/book/src/developer/logging.md
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:7]: Switched to end user, defined in https://cluster-api.sigs.k8s.io/user/personas.html#service-provider-end-userconsumer

File: docs/book/src/developer/logging.md
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:13]: it is intentionally must use

File: docs/book/src/developer/logging.md
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:31]: done

File: docs/book/src/developer/logging.md
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:133]: I have dropped this sentence and renamed the paragraph trade-offs
I agree that in some cases it could make sense to get closer to an IDE debugger.

File: docs/book/src/developer/logging.md
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:23]: done

**Comment by vincepri** [comment]
LGTM

File: docs/book/src/developer/logging.md
Comment (ykakarap) [inline docs/book/src/developer/logging.md:121]:
```suggestion
Ideally, in a future release of Cluster API we will switch to use 2 as a default verbosity (currently it is 0) for all the Cluster API
controllers as recommended by the Kubernetes guidelines.
```

This reads better IMHO. Up to you 🙂

**Review by ykakarap — COMMENTED** [review]
This is great overall! 🚀

File: docs/book/src/developer/logging.md
Comment (fabriziopandini) [reply docs/book/src/developer/logging.md:19]: https://github.com/kubernetes-sigs/cluster-api/pull/6969/files/91ccef7983b99f93bdaa30f8846f0c29e46dbbb3#diff-f519e7a9f2f1e96a5cbee2b82026a0de78a1836f889153e485e6581a7ad83e92 required to remove a similar note.
Also, we have specific pointers to Kubernetes guidelines all around the doc, so I don't think that having something super generic here adds much value.

Comment (ykakarap) [reply docs/book/src/developer/logging.md:19]: Should we just drop the `Upstream Alignment` section? I guess it should be fine since we are already referencing the upstream guidelines at various places in the document.

**Review by ykakarap — COMMENTED** [review]
LGTM pending squash.

**Comment by fabriziopandini** [comment]
As per 27th July office hours discussion lazy consensus set for EOW (29th of July)

**Review by elmiko — COMMENTED** [review]
LGTM

i generally like the guidelines here and picked up a few things from reading them. one thing that stood out to me is the advice to log before the action. i have a feeling this will cause many providers to re-examine their logging code, i know i'd like to go check the kubemark stuff!

**Comment by fabriziopandini** [comment]
Commit squashed
/hold cancel
as per community decision

**Review by chrischdi — COMMENTED** [review]
/lgtm

**Review by vincepri — APPROVED** [review]
/approve

**Comment by killianmuldoon** [comment]
/cherry-pick 1.2

To get this into the current version of the cluster API book

**Pull request merged**
